### PR TITLE
[build] Dropping Dependency on `nix`

### PR DESCRIPTION
**Pull request by ppenna**
## Description

This PR closes https://github.com/demikernel/demikernel/issues/402.

## Summary of Changes

- Dropped dependency on `nix` crate for Catnap LibOS.
- Dropped dependency on `nix` crate for Catcollar LibOS.

### CODE REVIEW

File: src/rust/catcollar/futures/accept.rs [hunk 1]
```diff
@@ -64,46 +68,63 @@ impl AcceptFuture {
...
 impl Future for AcceptFuture {
     type Output = Result<RawFd, Fail>;
 
     /// Polls the underlying accept operation.
     fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
-        let self_: &AcceptFuture = self.get_mut();
-        match socket::accept(self_.fd as i32) {
+        let self_: &mut AcceptFuture = self.get_mut();
+        match unsafe {
+            let mut address_len: libc::socklen_t = 0;
```
Comment (BrianZill) [inline src/rust/catcollar/futures/accept.rs:83]: Why is this zero?  Shouldn't this be sizeof sockaddr_in?  Assuming we actually want the address filled in, of course.

File: src/rust/catcollar/futures/accept.rs [hunk 2]
```diff
@@ -64,46 +68,63 @@ impl AcceptFuture {
...
-            Err(e) if e == Errno::EWOULDBLOCK || e == Errno::EAGAIN => {
-                ctx.waker().wake_by_ref();
-                Poll::Pending
-            },
-            // Operation failed.
-            Err(e) => {
-                warn!("failed to accept connection ({:?})", e);
-                Poll::Ready(Err(Fail::new(e as i32, "operation failed")))
+
+            // Operation not completed, thus parse errno to find out what happened.
```
Comment (BrianZill) [inline src/rust/catcollar/futures/accept.rs:113]: I would put all these comments inside the block to which they apply, as they describe state that is only true inside the block.

File: src/rust/catcollar/futures/connect.rs [hunk 3]
```diff
@@ -69,22 +71,34 @@ impl Future for ConnectFuture {
     /// Polls the underlying connect operation.
     fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
         let self_: &mut ConnectFuture = self.get_mut();
-        match socket::connect(self_.fd as i32, &self_.addr) {
+        match unsafe {
+            libc::connect(
+                self_.fd,
+                (&self_.sockaddr as *const libc::sockaddr_in) as *const libc::sockaddr,
+                mem::size_of_val(&self_.sockaddr) as u32,
```
Comment (BrianZill) [inline src/rust/catcollar/futures/connect.rs:78]: While this will work, since sizeof sockaddr >= sizeof sockaddr_in, this really should be the latter and not the former.

File: src/rust/catcollar/futures/connect.rs [hunk 4]
```diff
@@ -69,22 +71,34 @@ impl Future for ConnectFuture {
...
-                warn!("failed to establish connection ({:?})", errno);
-                Poll::Ready(Err(Fail::new(errno as i32, "operation failed")))
+
+            // Operation not completed, thus parse errno to find out what happened.
+            _ => {
+                let errno: libc::c_int = unsafe { *libc::__errno_location() };
+
+                // Operation in progress.
+                if errno == libc::EINPROGRESS || errno == libc::EALREADY {
+                    ctx.waker().wake_by_ref();
```
Comment (BrianZill) [inline src/rust/catcollar/futures/connect.rs:93]: Keep the trace! message?

File: src/rust/catnap/futures/accept.rs [hunk 5]
```diff
@@ -62,46 +66,63 @@ impl AcceptFuture {
...
 impl Future for AcceptFuture {
     type Output = Result<RawFd, Fail>;
 
     /// Polls the target [AcceptFuture].
     fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
-        let self_: &AcceptFuture = self.get_mut();
-        match socket::accept(self_.fd as i32) {
+        let self_: &mut AcceptFuture = self.get_mut();
+        match unsafe {
+            let mut address_len: libc::socklen_t = 0;
```
Comment (BrianZill) [inline src/rust/catnap/futures/accept.rs:81]: This one too.

File: src/rust/catnap/futures/connect.rs [hunk 6]
```diff
@@ -68,21 +70,34 @@ impl Future for ConnectFuture {
     /// Polls the target [ConnectFuture].
     fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
         let self_: &mut ConnectFuture = self.get_mut();
-        match socket::connect(self_.fd as i32, &self_.addr) {
+        match unsafe {
+            libc::connect(
+                self_.fd,
+                (&self_.sockaddr as *const libc::sockaddr_in) as *const libc::sockaddr,
+                mem::size_of_val(&self_.sockaddr) as u32,
```
Comment (BrianZill) [inline src/rust/catnap/futures/connect.rs:77]: Same comment as for the catcollar version.

File: src/rust/catnap/futures/pop.rs [hunk 7]
```diff
@@ -78,34 +81,40 @@ impl Future for PopFuture {
     /// Polls the target [PopFuture].
     fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
         let self_: &mut PopFuture = self.get_mut();
-        let mut bytes: [u8; POP_SIZE] = [0; POP_SIZE];
-        match socket::recvfrom::<SockaddrStorage>(self_.fd, &mut bytes[..]) {
+        match unsafe {
+            let mut addrlen: libc::socklen_t = 0;
```
Comment (BrianZill) [inline src/rust/catnap/futures/pop.rs:85]: Another one.

File: src/rust/catnap/futures/pop.rs [hunk 8]
```diff
@@ -78,34 +81,40 @@ impl Future for PopFuture {
...
+            libc::recvfrom(
+                self_.fd,
+                (self_.buf.as_mut_ptr() as *mut u8) as *mut libc::c_void,
+                self_.buf.len(),
+                libc::MSG_DONTWAIT,
```
Comment (BrianZill) [inline src/rust/catnap/futures/pop.rs:90]: Why this change?  Shouldn't this fd already be set_nonblock?

File: src/rust/catnap/futures/push.rs [hunk 9]
```diff
@@ -66,21 +63,37 @@ impl Future for PushFuture {
     /// Polls the target [PushFuture].
     fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
         let self_: &mut PushFuture = self.get_mut();
-        match socket::send(self_.fd, &self_.buf[..], socket::MsgFlags::empty()) {
+        match unsafe {
+            libc::sendto(
```
Comment (BrianZill) [inline src/rust/catnap/futures/push.rs:67]: Why sendto instead of send here?

File: src/rust/catcollar/mod.rs [hunk 10]
```diff
@@ -162,9 +160,21 @@ impl CatcollarLibOS {
...
             Some(&fd) => {
-                let addr: SockaddrStorage = parse_addr(local);
-                socket::bind(fd, &addr).unwrap();
-                Ok(())
+                let sockaddr: libc::sockaddr_in = linux::socketaddrv4_to_sockaddr_in(&local);
+                match unsafe {
+                    libc::bind(
+                        fd,
+                        (&sockaddr as *const libc::sockaddr_in) as *const libc::sockaddr,
+                        mem::size_of_val(&sockaddr) as u32,
```
Comment (BrianZill) [inline src/rust/catcollar/mod.rs:168]: Should be sizeof sockaddr_in.

File: src/rust/catcollar/futures/connect.rs (hunk 3 above)
Comment (BrianZill) [reply src/rust/catcollar/futures/connect.rs:78]: Nevermind, I see `sockaddr` is a `libc::sockaddr_in`, so this is ok.

File: src/rust/catnap/futures/connect.rs [hunk 11]
```diff
@@ -68,21 +70,34 @@ impl Future for ConnectFuture {
     /// Polls the target [ConnectFuture].
     fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
         let self_: &mut ConnectFuture = self.get_mut();
-        match socket::connect(self_.fd as i32, &self_.addr) {
+        match unsafe {
+            libc::connect(
```
Comment (iyzhang) [inline src/rust/catnap/futures/connect.rs:74]: I noticed that every libc invokation requires an unsafe block. Is that because these functions are unsafe? Are the nix ones safe somehow?

File: src/rust/catnap/futures/pop.rs [hunk 12]
```diff
@@ -31,34 +25,43 @@ use ::std::{
...
 
 /// Associate Functions for Pop Operation Descriptors
 impl PopFuture {
     /// Creates a descriptor for a pop operation.
     pub fn new(qd: QDesc, fd: RawFd) -> Self {
-        Self { qd, fd }
+        Self {
+            qd,
+            fd,
+            sockaddr: unsafe { mem::zeroed() },
```
Comment (iyzhang) [inline src/rust/catnap/futures/pop.rs:62]: Is this unsafe block necessary?

File: src/rust/catnap/futures/mod.rs [hunk 13]
```diff
@@ -1,19 +1,19 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
-//==============================================================================
+//======================================================================================================================
```
Comment (iyzhang) [inline src/rust/catnap/futures/mod.rs:4]: Not sure if this is intentional, but this file was only changed to extend the comment bars. I don't mind having style fixes in this PR, but it should probably have its own commit.

File: src/rust/catcollar/futures/mod.rs [hunk 14]
```diff
@@ -1,19 +1,19 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
-//==============================================================================
+//======================================================================================================================
```
Comment (iyzhang) [inline src/rust/catcollar/futures/mod.rs:4]: Same on this file. We should be sure to separate style changes from code changes.

File: src/rust/catcollar/futures/accept.rs (hunk 1 above)
Comment (ppenna) [reply src/rust/catcollar/futures/accept.rs:83]: You are correct. I have fixed this. Thanks for checking this.

File: src/rust/catnap/futures/accept.rs (hunk 5 above)
Comment (ppenna) [reply src/rust/catnap/futures/accept.rs:81]: You are correct. I have fixed this. Thanks for checking this.

File: src/rust/catnap/futures/pop.rs (hunk 7 above)
Comment (ppenna) [reply src/rust/catnap/futures/pop.rs:85]: You are correct. I have fixed this. Thanks for checking this.

File: src/rust/catcollar/futures/accept.rs (hunk 2 above)
Comment (ppenna) [reply src/rust/catcollar/futures/accept.rs:113]: Sure. I moved them in there.

File: src/rust/catcollar/futures/connect.rs (hunk 3 above)
Comment (ppenna) [reply src/rust/catcollar/futures/connect.rs:78]: Thanks

File: src/rust/catcollar/futures/connect.rs (hunk 4 above)
Comment (ppenna) [reply src/rust/catcollar/futures/connect.rs:93]: Sure! Thanks for noting.

File: src/rust/catnap/futures/pop.rs (hunk 8 above)
Comment (ppenna) [reply src/rust/catnap/futures/pop.rs:90]: When creating the socket the `set_nonblock()` call may not succeed. In that case the socket is still created, but it would be a blocking one. By supplying the `libc::MSG_DONTWAIT` we are doing our best to make any asynchronous operation in a blocking socket.

File: src/rust/catnap/futures/push.rs (hunk 9 above)
Comment (ppenna) [reply src/rust/catnap/futures/push.rs:67]: For two reasons:
1. To be able to set the `MSG_DONTWAIT` flag (see my reply from above)
2. To intentionally align up the logic of `PushFuture` and `PopFuture`.

For (2) I do have a change on my queue that unifies the logic of both co-routines.

File: src/rust/catnap/futures/connect.rs (hunk 11 above)
Comment (ppenna) [reply src/rust/catnap/futures/connect.rs:74]: Yes, invoking `libc` functions is unsafe. Functions from nix calls wrap around and provide a "safe" interface.

File: src/rust/catnap/futures/pop.rs (hunk 12 above)
Comment (ppenna) [reply src/rust/catnap/futures/pop.rs:62]: Yes it is: https://doc.rust-lang.org/std/mem/fn.zeroed.html

File: src/rust/catnap/futures/mod.rs (hunk 13 above)
Comment (ppenna) [reply src/rust/catnap/futures/mod.rs:4]: I was following our agreement that if something would be off the "coding style" it could be cleaned up. Anyways, I have reverted these changes.

File: src/rust/catcollar/futures/mod.rs (hunk 14 above)
Comment (ppenna) [reply src/rust/catcollar/futures/mod.rs:4]: I was following our agreement that if something would be off the "coding style" it could be cleaned up. Anyways, I have reverted these changes.

File: src/rust/catnap/futures/mod.rs (hunk 13 above)
Comment (iyzhang) [reply src/rust/catnap/futures/mod.rs]: Yes, that is fine but I just wanted a separate commit in the PR for the style changes.

File: src/rust/catcollar/futures/accept.rs (hunk 1 above)
Comment (BrianZill) [reply src/rust/catcollar/futures/accept.rs]: Great, thanks!

File: src/rust/catcollar/futures/accept.rs (hunk 2 above)
Comment (BrianZill) [reply src/rust/catcollar/futures/accept.rs:113]: Thanks!

File: src/rust/catnap/futures/pop.rs (hunk 8 above)
Comment (BrianZill) [reply src/rust/catnap/futures/pop.rs:90]: Okay.

File: src/rust/catnap/futures/push.rs (hunk 9 above)
Comment (BrianZill) [reply src/rust/catnap/futures/push.rs:67]: Okay.

**Review by BrianZill — APPROVED** [review]
Looks good now.  Thanks for removing this external dependency!
